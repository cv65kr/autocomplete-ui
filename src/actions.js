/**
 * Actions
 */
export const actions = store => ({
    /**
     * Search Action
     * @param state
     * @param queryText
     */
    searchAction(state, queryText) {
        let query = state
            .client
            .query
            .create(queryText)
            .enableHighlights()
        ;

        state
            .client
            .search(query, (data, error) => {
                if (error) {
                    store.setState({error});
                    return;
                }

                store.setState({data});
            })
    },
});