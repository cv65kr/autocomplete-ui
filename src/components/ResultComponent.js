import {h} from "preact"
import { connect } from 'unistore/preact'
import groupBy from 'lodash/groupBy'
import {
    buildPoweredBy,
    createDatasetByItemTypeAndSetActiveIndex,
    renderTemplate
} from "./helpers";

const defaultHtmlAttributes = {
    id: "apisearch-listbox",
    className: "as-result",
    role: "listbox",
    tabIndex: "-1"
};

/**
 * Result Component
 *
 * @param datasets
 * @param poweredBy
 * @param resultBoxOpen
 * @param currentCursorIndex
 * @param items
 * @returns {XML}
 * @constructor
 */
export const ResultComponent = ({
   datasets,
   poweredBy,
   resultBoxOpen,
   currentCursorIndex,
   items
}) => {
    /**
     * No results
     */
    if (false === resultBoxOpen) {
        return <div
            {...defaultHtmlAttributes}
            style={{display: 'none'}}
        />;
    }

    let filteredItemsByType = createDatasetByItemTypeAndSetActiveIndex(
        items,
        currentCursorIndex
    );
    let filteredDatasetsByType = groupBy(datasets, 'type');

    /**
     * Results set
     */
    return (
        <div
            {...defaultHtmlAttributes}
        >
            {Object.keys(filteredItemsByType).map(type =>
                <div
                    className={`as-result__dataset as-result__dataset--${type}`}
                >
                    {(filteredDatasetsByType[type][0].template.header)
                        ? <div
                            className="as-result__datasetHeader"
                            dangerouslySetInnerHTML={
                                renderTemplate(
                                    filteredDatasetsByType[type][0].template.header,
                                    null
                                )
                            }
                        />
                        : null
                    }

                    <ul
                        className="as-result__datasetItemsList"
                    >
                        {filteredItemsByType[type].map((item, index) =>
                            <li
                                data-index={item.navigationIndex}
                                className={`as-result__datasetItem${
                                    item.isActive
                                        ? ' as-result__datasetItem--active'
                                        : ''
                                    }`}
                                dangerouslySetInnerHTML={
                                    renderTemplate(
                                        filteredDatasetsByType[type][0].template.item,
                                        item
                                    )
                                }
                            />
                        )}
                    </ul>
                </div>
            )}

            {buildPoweredBy(poweredBy)}
        </div>
    )
};

/**
 * Result connector
 */
const ConnectedResult = connect('datasets, poweredBy, resultBoxOpen, currentCursorIndex, items')(ResultComponent);

export default ConnectedResult;