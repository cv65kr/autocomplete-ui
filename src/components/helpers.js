import {h} from "preact";
import {groupBy, sortBy} from "lodash";
import Hogan from "hogan.js";


/**
 * Compile template
 * @param template
 * @param data
 */
export const renderTemplate = (template, data) => {
    let compiledTemplate = Hogan.compile(template);

    return {
        __html: compiledTemplate.render(data)
    }
};

/**
 * Set active index
 *
 * @param items
 * @param currentCursorIndex
 * @returns {{}}
 */
export const createDatasetByItemTypeAndSetActiveIndex = (
    items,
    currentCursorIndex
) => {
    let sortedItemsByType = sortBy(items, 'uuid.type');
    let itemsWithActiveIndex = sortedItemsByType.map((item, index) =>
        ({
            ...item,
            navigationIndex: (index + 1),
            isActive: ((index + 1) === currentCursorIndex)
        })
    );

    return groupBy(itemsWithActiveIndex, 'uuid.type');
};

/**
 * Build the poweredBy icon
 * It can be a boolean:
 *   if (false) do now show a poweredBy
 *   if (true) assign the default poweredBy
 *   else render a custom poweredBy template string
 *
 * @param poweredBy
 * @returns {XML}
 */
export const buildPoweredBy = (poweredBy = false) => {
    if (false === poweredBy) return null;

    if (true === poweredBy) {
        poweredBy = `<a href="http://apisearch.io">
            <img 
                src="http://apisearch.io/assets/media/powered-by.svg" 
                alt="Powered by Apisearch" 
                width="100px"
            />
        </a>`;
    }

    return <div
        className="as-poweredBy"
        dangerouslySetInnerHTML={
            renderTemplate(poweredBy, null)
        }
    />
};