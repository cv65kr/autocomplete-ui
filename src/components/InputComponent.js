import {h} from "preact";
import {actions} from "../actions";
import { connect } from 'unistore/preact'

/**
 * Input Component
 *
 * @param htmlNodeInheritProps
 * @param searchAction
 * @param keyDownAction
 * @param focusOutAction
 * @constructor
 */
export const InputComponent = ({
   /** component props */
   htmlNodeInheritProps,

   /** actions */
   searchAction,
   keyDownAction,
   focusOutAction
}) => <input
    {...htmlNodeInheritProps}
    autocomplete="off"
    spellCheck="false"
    role="combobox"
    aria-autocomplete="list"
    aria-expanded="false"
    aria-owns="apisearch-listbox"
    data-search="Apisearch-autocomplete"

    onInput={searchAction}
    onKeyDown={keyDownAction}
    onBlur={focusOutAction}
/>;

/**
 * Input connector
 */
const ConnectedInput = connect('', actions)(InputComponent);

export default ConnectedInput;