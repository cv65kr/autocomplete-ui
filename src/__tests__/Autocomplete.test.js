import {h} from 'preact';
import {render} from 'preact'
import {actions} from "./__mocks__/actions.mock";
import createStore from "unistore";
import { connect, Provider } from 'unistore/preact'
import {InputComponent} from '../components/InputComponent';
import {ResultComponent} from '../components/ResultComponent';

const datasets = [
    {
        type: 'mock',
        template: {
            header: '<h3>Mock Dataset</h3>',
            item: '<p>{{metadata.name}}</p>'
        }
    }
];

describe('Autocomplete', () => {
    let initialState = {
        poweredBy: false,
        datasets,
        itemsPerResult: 10,
        startSearchOn: 0,
        queryText: "",
        resultBoxOpen: false,
        currentCursorIndex: 0,
        items: [],
        total_hits: 0
    };

    let onInputEvent = new Event('input', {
        'bubbles': true,
        'cancelable': true
    });
    let onKeydownEvent = new KeyboardEvent('keydown', {
        'bubbles': true,
        'cancelable': true
    });
    let onBlurEvent = new KeyboardEvent('blur', {
        'bubbles': true,
        'cancelable': true
    });

    it('should render', () => {
        const ConnectedInput = connect('', actions)(InputComponent);
        const ConnectedResult = connect('datasets, poweredBy, resultBoxOpen, currentCursorIndex, items', actions)(ResultComponent);
        const store = createStore(initialState);

        /**
         * Render autocomplete
         */
        let root = document.createElement('div');
        render(
            <Provider store={store}>
                <ConnectedInput htmlNodeInheritProps={{
                    id: 'apisearch-input-id',
                    class: 'some-class-name'
                }}/>
                <ConnectedResult />
            </Provider>,
            root
        );

        /**
         * Expect initial store state
         */
        expect(store.getState()).toMatchSnapshot();

        /**
         * Searching on input --> "some query text"
         * Expect results on store state
         */
        root.firstElementChild.value = 'some query text';
        root.firstElementChild.dispatchEvent(onInputEvent);
        expect(store.getState()).toMatchSnapshot();

        /**
         * KeyDown action to change the cursor index
         */
        root.firstElementChild.dispatchEvent(onKeydownEvent);
        expect(store.getState()).toMatchSnapshot();

        /**
         * Focus out action
         */
        root.firstElementChild.dispatchEvent(onBlurEvent);
        expect(store.getState()).toMatchSnapshot();
    });
});
