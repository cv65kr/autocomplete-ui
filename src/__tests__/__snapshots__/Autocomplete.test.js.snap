// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete should render 1`] = `
Object {
  "currentCursorIndex": 0,
  "datasets": Array [
    Object {
      "template": Object {
        "header": "<h3>Mock Dataset</h3>",
        "item": "<p>{{metadata.name}}</p>",
      },
      "type": "mock",
    },
  ],
  "items": Array [],
  "itemsPerResult": 10,
  "poweredBy": false,
  "queryText": "",
  "resultBoxOpen": false,
  "startSearchOn": 0,
  "total_hits": 0,
}
`;

exports[`Autocomplete should render 2`] = `
Object {
  "currentCursorIndex": 0,
  "datasets": Array [
    Object {
      "template": Object {
        "header": "<h3>Mock Dataset</h3>",
        "item": "<p>{{metadata.name}}</p>",
      },
      "type": "mock",
    },
  ],
  "items": Array [
    Object {
      "metadata": Object {
        "name": "This is a mocked item",
      },
      "uuid": Object {
        "id": 1,
        "type": "mock",
      },
    },
    Object {
      "metadata": Object {
        "name": "This is a mocked item 2",
      },
      "uuid": Object {
        "id": 2,
        "type": "mock",
      },
    },
    Object {
      "metadata": Object {
        "name": "This is a mocked item 3",
      },
      "uuid": Object {
        "id": 3,
        "type": "mock",
      },
    },
  ],
  "itemsPerResult": 10,
  "poweredBy": false,
  "queryText": "some query text",
  "resultBoxOpen": true,
  "startSearchOn": 0,
  "total_hits": 3,
}
`;

exports[`Autocomplete should render 3`] = `
Object {
  "currentCursorIndex": 1,
  "datasets": Array [
    Object {
      "template": Object {
        "header": "<h3>Mock Dataset</h3>",
        "item": "<p>{{metadata.name}}</p>",
      },
      "type": "mock",
    },
  ],
  "items": Array [
    Object {
      "metadata": Object {
        "name": "This is a mocked item",
      },
      "uuid": Object {
        "id": 1,
        "type": "mock",
      },
    },
    Object {
      "metadata": Object {
        "name": "This is a mocked item 2",
      },
      "uuid": Object {
        "id": 2,
        "type": "mock",
      },
    },
    Object {
      "metadata": Object {
        "name": "This is a mocked item 3",
      },
      "uuid": Object {
        "id": 3,
        "type": "mock",
      },
    },
  ],
  "itemsPerResult": 10,
  "poweredBy": false,
  "queryText": "some query text",
  "resultBoxOpen": true,
  "startSearchOn": 0,
  "total_hits": 3,
}
`;

exports[`Autocomplete should render 4`] = `
Object {
  "currentCursorIndex": 1,
  "datasets": Array [
    Object {
      "template": Object {
        "header": "<h3>Mock Dataset</h3>",
        "item": "<p>{{metadata.name}}</p>",
      },
      "type": "mock",
    },
  ],
  "items": Array [
    Object {
      "metadata": Object {
        "name": "This is a mocked item",
      },
      "uuid": Object {
        "id": 1,
        "type": "mock",
      },
    },
    Object {
      "metadata": Object {
        "name": "This is a mocked item 2",
      },
      "uuid": Object {
        "id": 2,
        "type": "mock",
      },
    },
    Object {
      "metadata": Object {
        "name": "This is a mocked item 3",
      },
      "uuid": Object {
        "id": 3,
        "type": "mock",
      },
    },
  ],
  "itemsPerResult": 10,
  "poweredBy": false,
  "queryText": "some query text",
  "resultBoxOpen": false,
  "startSearchOn": 0,
  "total_hits": 3,
}
`;
